src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc

```diff
@@ -164,7 +163,11 @@ AresDnsResolver::~AresDnsResolver() {
   if (resolved_result_ != nullptr) {
     grpc_channel_args_destroy(resolved_result_);
   }
-  grpc_pollset_set_destroy(interested_parties_);
+  if (next_completion_ != nullptr) {
+    GRPC_CLOSURE_SCHED(next_completion_, GRPC_ERROR_CREATE_FROM_STATIC_STRING(
```

Comment:
I don't think this should be done in the dtor.  Instead, it should be triggered from `OnResolvedLocked()`.

Reply:
I'm stuck on this one. Relying on `OnResolvedLocked` I think isn't safe because it thus only schedules `next_completion_` in the case that `ShutdownLocked` was called with a pending query... which isn't true e.g. when the re-resolution backoff timer is preventing the next query from starting.

Reply:
I still have a concern here about the way that the ref from resolver to client channel is dropped. What's to prevent the following race from happening:

1) client channel [orphans its resolver](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.cc#L635), and so schedules resolver's `ShutdownLocked`
2) `ShutdownLocked` closure hasn't <i>ran</i> yet. Meanwhile a pending resolution completes and resolver schedules its `next_completion_` without an error
3) client channel's `next_completion_` callback is invoked and client channel's resolver `== nullptr` so it proceeds to [shut down](https://github.com/grpc/grpc/blob/master/src/core/ext/filters/client_channel/client_channel.cc#L515). Client channel now drops it's resolver's channel stack ref
4) Resolver's `ShutdownLocked` closure is still scheduled but still hasn't ran. Meanwhile: re-resolution timer goes off and we start resolving - at this point we will access channel's pollset set in an unsafe way.

Reply:
Because of the issue about the current single resolver -> client ref potentially not being enough to keep the channel's pollset set alive while the resolver might use it, I ended up adding the extra ref from resolver -> channel... which I believe ends up covering all cases here.

PTAL, if this now looks ok